pkg/webhook/route/route.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package route
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/webhook/statefulset"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	"k8s.io/api/admission/v1beta1"
+)
+
+// admitFunc is the type we use for all of our validators
+type admitFunc func(v1beta1.AdmissionReview) *v1beta1.AdmissionResponse
+
+// marshal responseAdmissionReview and send back
+func marshalAndWrite(response v1beta1.AdmissionReview, w http.ResponseWriter) {
+
+	respBytes, err := json.Marshal(response)
+	if err != nil {
+		glog.Errorf("%v", err)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		glog.Errorf("%v", err)
+	}
+
+}
+
+// serve handles the http portion of a request prior to handing to an admit
+// function
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	var body []byte
+	var contentType string
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+	deserializer := util.GetCodec()
+
+	// The AdmissionReview that will be returned
+	if r.Body != nil {
```

Comment:
```
	if r.Body == nil {
		err := errors.New("request body is nil!")
		responseAdmissionReview.Response = util.ARFail(err)
		marshalAndWrite(responseAdmissionReview, w)
		return
        }

        if data, err := ioutil.ReadAll(r.Body); err != nil {
          responseAdmissionReview.Response = util.ARFail(err)
          marshalAndWrite(responseAdmissionReview, w)
          return
        }

        body = data
```

Reply:
data, err := ioutil.ReadAll(r.Body)
        if err != nil {
          responseAdmissionReview.Response = util.ARFail(err)
          marshalAndWrite(responseAdmissionReview, w)
          return
        }

        body = data

       done like this